modules/openxr/openxr_api.cpp

```diff
@@ -0,0 +1,2162 @@
+	}
+}
+
+bool OpenXRAPI::create_swapchain(int64_t p_swapchain_format, uint32_t p_width, uint32_t p_height, uint32_t p_sample_count, uint32_t p_array_size, XrSwapchain &r_swapchain, void **r_swapchain_graphics_data) {
+	ERR_FAIL_COND_V(session == XR_NULL_HANDLE, false);
+	ERR_FAIL_NULL_V(graphics_extension, false);
+
+	XrResult result;
+
+	void *next_pointer = nullptr;
+	for (OpenXRExtensionWrapper *wrapper : registered_extension_wrappers) {
+		next_pointer = wrapper->set_swapchain_create_info_and_get_next_pointer(next_pointer);
+	}
+
+	XrSwapchainCreateInfo swapchain_create_info = {
+		XR_TYPE_SWAPCHAIN_CREATE_INFO, // type
+		next_pointer, // next
+		0, // createFlags
+		XR_SWAPCHAIN_USAGE_SAMPLED_BIT | XR_SWAPCHAIN_USAGE_COLOR_ATTACHMENT_BIT, // usageFlags
+		p_swapchain_format, // format
+		p_sample_count, // sampleCount
+		p_width, // width
+		p_height, // height
+		1, // faceCount
+		p_array_size, // arraySize
+		1 // mipCount
+	};
+
+	XrSwapchain new_swapchain;
+	result = xrCreateSwapchain(session, &swapchain_create_info, &new_swapchain);
+	if (XR_FAILED(result)) {
+		print_line("OpenXR: Failed to get swapchain [", get_error_string(result), "]");
+		return false;
+	}
+
+	if (!graphics_extension->get_swapchain_image_data(new_swapchain, p_swapchain_format, p_width, p_height, p_array_size, p_array_size, r_swapchain_graphics_data)) {
+		xrDestroySwapchain(new_swapchain);
+		return false;
+	}
+
+	r_swapchain = new_swapchain;
+
+	return true;
+}
+
+bool OpenXRAPI::on_state_idle() {
+#ifdef DEBUG
+	print_line("On state idle");
+#endif
+
+	for (OpenXRExtensionWrapper *wrapper : registered_extension_wrappers) {
+		wrapper->on_state_idle();
+	}
+
+	return true;
+}
+
+bool OpenXRAPI::on_state_ready() {
+#ifdef DEBUG
+	print_line("On state ready");
+#endif
+
+	// begin session
+	XrSessionBeginInfo session_begin_info = {
+		XR_TYPE_SESSION_BEGIN_INFO, // type
+		nullptr, // next
+		view_configuration // primaryViewConfigurationType
+	};
+
+	XrResult result = xrBeginSession(session, &session_begin_info);
+	if (XR_FAILED(result)) {
+		print_line("OpenXR: Failed to begin session [", get_error_string(result), "]");
+		return false;
+	}
+
+	// This is when we create our swapchain, this can be a "long" time after Godot finishes, we can deal with this for now
+	// but once we want to provide Viewports for additional layers where OpenXR requires us to create further swapchains,
+	// we'll be creating those viewport WAY before we reach this point.
+	// We may need to implement a wait in our init in main.cpp polling our events until the session is ready.
+	// That will be very very ugly
+	// The other possibility is to create a separate OpenXRViewport type specifically for this goal as part of our OpenXR module
+
+	if (!create_main_swapchain()) {
+		return false;
+	}
+
+	// we're running
+	running = true;
+
+	for (OpenXRExtensionWrapper *wrapper : registered_extension_wrappers) {
+		wrapper->on_state_ready();
+	}
+
+	// TODO emit signal
```

Comment:
Is this to be added in a follow-up PR?

Reply:
Yes, I wanted to get the base line working first, there are a bunch of things we're doing in the plugin that either need to still be ported but are waiting on something (like Meta adopting the Khronos loader) and things that need to change.

But seeing this PR already covers a whole bunch of things, plus that this will allow us to divide some of the work